Added AKS Audit Logs Rule

ArmaanMcleod opened the pull request:
## PR Summary

Fix #882

Added `Azure.AKS.AuditLogs` rule for the following resource types:

- Microsoft.ContainerService/managedClusters

It checks if one of `kube-audit`/`kub-audit-admin` and `guard` are enabled.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

ArmaanMcleod commented:
No clue why powershell 5.1 is failing. Will have a look tomorrow.

BernieWhite commented on tests/PSRule.Rules.Azure.Tests/Cmdlet.Common.Tests.ps1:
```diff
@@ -531,28 +531,49 @@ InModuleScope -ModuleName 'PSRule.Rules.Azure' {
+
+            Mock -CommandName 'Get-AzResource' -MockWith {
+                return @(
+                    [PSCustomObject]@{
```
If not too much work, may lets populate with a basic `microsoft.insights/diagnosticSettings` resource, just update ids and add empty logs array under properties instead of shared key.

ArmaanMcleod replied:
Definetely not too much work. Was thinking of doing that yesterday. Changed it to a be a diagnostSettings resource instead 🙂

BernieWhite reviewed (approved):
@ArmaanMcleod Suggested wrapping array blocks should address build issue or something similar.

Great work, killing it.

ArmaanMcleod commented on src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1:
```diff
@@ -223,6 +223,28 @@ Rule 'Azure.AKS.ContainerInsights' -Type 'Microsoft.ContainerService/managedClus
+Rule 'Azure.AKS.AuditLogs' -Type 'Microsoft.ContainerService/managedClusters' -Tag @{ release = 'GA'; ruleSet = '2021_09'; } {
+    $diagnosticLogs = @(GetSubResources -ResourceType 'Microsoft.Insights/diagnosticSettings', 'Microsoft.ContainerService/managedClusters/providers/diagnosticSettings');
+
+    $null -ne $diagnosticLogs -and $diagnosticLogs.Length -gt 0;
```
Ah great, much cleaner. Thanks mate.

ArmaanMcleod commented on src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1:
```diff
@@ -223,6 +223,28 @@ Rule 'Azure.AKS.ContainerInsights' -Type 'Microsoft.ContainerService/managedClus
+    $null -ne $diagnosticLogs -and $diagnosticLogs.Length -gt 0;
+
+    foreach ($setting in $diagnosticLogs) {
+        $kubeAuditEnabledLog = $setting.Properties.logs | Where-Object {
```
Ah that would fix it. Always forget that causes issues in PS5.1.

ArmaanMcleod commented:
> @ArmaanMcleod Suggested wrapping array blocks should address build issue or something similar.
> 
> Great work, killing it.

Thanks @BernieWhite, that definetely solved the issue. I always seem to forget thats needed in PS5.1, so used to PS7+ and not having to worry about that 😄 . 

Cleaned up code a bit with your suggestions and modified test to use diagnostic resource. Definetely looks much better now 🙂

ArmaanMcleod commented:
/azp run

ArmaanMcleod commented:
Tried rerunning pipeline again to see if Bicep tests pass, seems like it didn't do nything 😄 .

ArmaanMcleod commented:
/azp run

ArmaanMcleod commented:
@BernieWhite All good to go, let me know if your fine with me merging this 🙂

BernieWhite reviewed (approved):
@ArmaanMcleod Nice work. All good to merge.
